src/core/server/elasticsearch/version_check/ensure_es_version.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * ES and Kibana versions are locked, so Kibana should require that ES has the same version as
+ * that defined in Kibana's package.json.
+ */
+
+import { get } from 'lodash';
+import { coerce } from 'semver';
+import { defer, timer, iif, throwError } from 'rxjs';
+import { retryWhen, concatMap, map } from 'rxjs/operators';
+import { isEsCompatibleWithKibana } from './is_es_compatible_with_kibana';
+import { Logger } from '../../logging';
+import { APICaller } from '..';
+
+export interface EnsureVersionOptions {
+  callWithInternalUser: APICaller;
+  log: Logger;
+  kibanaVersion: string;
+  ignoreVersionMismatch: boolean;
+  healthCheckDelay: number;
+}
+
+export interface NodesInfo {
+  nodes: {
+    [key: string]: NodeInfo;
+  };
+}
+
+interface NodeInfo {
+  version: string;
+  ip: string;
+  http: {
+    'publish.address': string;
+  };
+}
+
+function getHumanizedNodeName(node: NodeInfo) {
+  const publishAddress = get(node, 'http.publish_address')
+    ? get(node, 'http.publish_address') + ' '
+    : '';
+  return 'v' + node.version + ' @ ' + publishAddress + '(' + node.ip + ')';
+}
+
+export function ensureEsVersionMatch(
+  nodesInfo: NodesInfo,
+  log: Logger,
+  kibanaVersion: string,
+  ignoreVersionMismatch: boolean
+) {
+  // Aggregate incompatible ES nodes.
+  const incompatibleNodes = Object.keys(nodesInfo.nodes)
+    .map(key => nodesInfo.nodes[key])
+    .filter(node => !isEsCompatibleWithKibana(node.version, kibanaVersion))
+    .map(getHumanizedNodeName);
+
+  if (!ignoreVersionMismatch && incompatibleNodes.length > 0) {
+    throw new Error(
+      `This version of Kibana requires Elasticsearch v${kibanaVersion} on all nodes. I found ` +
+        `the following incompatible nodes in your cluster: ${incompatibleNodes.join(', ')}`
```

Comment:
NIT: first person in system errors?

Reply:
We don't have a good versioning/compatibility strategy with logs, but operations people might rely on logs as a kind of API to detect/alert on common failures. So I'm hesitant to change this in a minor, but I agree that the message should be rewritten.